File: app/src/main/java/com/example/android/architecture/scenes/post/PostFragment.java
package com.example.android.architecture.scenes.post;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.widget.DefaultItemAnimator;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import com.example.android.architecture.R;
import com.example.android.architecture.models.Post;
import java.util.ArrayList;
import java.util.List;

/**
 * Created by YaoHaitao on 2018/3/9.
 */

public class PostFragment extends Fragment implements PostContract.View {

  //region 组件声明
  private RecyclerView mRecyclerView;
  private TextView mNoDataTextView;
  //endregion

  //region 私有变量声明
  private PostContract.Presenter mPresenter;
  private PostAdapter mPostAdapter;
  //endregion

  //region 初始化
  public PostFragment() {
    super();
  }

  public static PostFragment newInstance() {
    return new PostFragment();
  }
  //endregion

  //region View 生命周期
  @Override
  public void onCreate(@Nullable Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);

    mPostAdapter = new PostAdapter(new ArrayList<Post>());
  }

  @Nullable
  @Override
  public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
      @Nullable Bundle savedInstanceState) {
    View root = inflater.inflate(R.layout.post_fragment, container, false);

    mRecyclerView = root.findViewById(R.id.rv_post);
    mRecyclerView.setLayoutManager(new LinearLayoutManager(getActivity()));
    mRecyclerView.setItemAnimator(new DefaultItemAnimator());
    mRecyclerView.setAdapter(mPostAdapter);

    mNoDataTextView = root.findViewById(R.id.tv_post_empty);

    return root;
  }

  @Override
  public void onResume() {
    super.onResume();

    mPresenter.start();
  }
  //endregion

  //region View Interface
  @Override
  public void setPresenter(PostContract.Presenter presenter) {
    this.mPresenter = presenter;
  }

  @Override
  public void showPosts(List<Post> posts) {
    mPostAdapter.replaceData(posts);

    mRecyclerView.setVisibility(View.VISIBLE);
    mNoDataTextView.setVisibility(View.GONE);
  }

  @Override
  public void showNoData() {
    mRecyclerView.setVisibility(View.GONE);
    mNoDataTextView.setVisibility(View.VISIBLE);
  }

  @Override
  public void toPostDetail(Post post) {
    // TODO
  }
  //endregion

  //region Recycler View
  private class PostHolder extends RecyclerView.ViewHolder {

    private Post mPost;

    private TextView mTitleTextView;
    private TextView mContentTextVIew;

    PostHolder(LayoutInflater inflater, ViewGroup parent) {
      super(inflater.inflate(R.layout.post_recycler_item_fragment, parent, false));

      mTitleTextView = itemView.findViewById(R.id.tv_post_item_title);
      mContentTextVIew = itemView.findViewById(R.id.tv_post_item_content);

      mTitleTextView.setText("");
      mContentTextVIew.setText("");

      itemView.setOnClickListener(
          view -> mPresenter.toPostDetail(mPost)
      );
    }

    void binding(Post post) {
      mPost = post;

      mTitleTextView.setText(post.getTitle());
      mContentTextVIew.setText(post.getContent());
    }
  }

  private class PostAdapter extends RecyclerView.Adapter<PostHolder> {

    private List<Post> mPosts;

    PostAdapter(List<Post> posts) {
      setPosts(posts);
    }

    void replaceData(List<Post> posts) {
      setPosts(posts);
      notifyDataSetChanged();
    }

    private void setPosts(List<Post> posts) {
      mPosts = posts;
    }

    @NonNull
    @Override
    public PostHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
      LayoutInflater inflater = LayoutInflater.from(getActivity());
      return new PostHolder(inflater, parent);
    }

    @Override
    public void onBindViewHolder(@NonNull PostHolder holder, int position) {
      Post post = mPosts.get(position);
      holder.binding(post);
    }

    @Override
    public int getItemCount() {
      return mPosts.size();
    }
  }
  //endregion
}

File: app/src/main/java/com/example/android/architecture/scenes/post/PostFragment.kt
package com.example.android.architecture.scenes.post

import android.os.Bundle
import android.support.v4.app.DialogFragment
import android.support.v4.app.Fragment
import android.support.v7.widget.DefaultItemAnimator
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.RecyclerView.Adapter
import android.support.v7.widget.RecyclerView.ViewHolder
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Switch
import android.widget.TextView
import com.example.android.architecture.R
import com.example.android.architecture.common.Constants
import com.example.android.architecture.models.Post
import com.example.android.architecture.scenes.post.PostContract.Presenter
import com.example.android.architecture.scenes.postdetail.EXTRA_POST_ID
import com.example.android.architecture.scenes.postdetail.PostDetailActivity
import com.example.android.architecture.utils.ActivityUtils

/**
 * ポストのフラグメント
 */
class PostFragment : Fragment(), PostContract.View {

  override lateinit var presenter: Presenter

  private lateinit var recyclerView: RecyclerView
  private lateinit var noDataTextView: TextView
  private lateinit var fingerprintSwitch: Switch

  private lateinit var postAdapter: PostAdapter

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)

    postAdapter = PostAdapter(ArrayList())
  }

  override fun onCreateView(
    inflater: LayoutInflater,
    container: ViewGroup?,
    savedInstanceState: Bundle?
  ): View? {
    val root: View = inflater.inflate(R.layout.post_fragment, container, false)

    recyclerView = root.findViewById<RecyclerView>(R.id.rv_post).apply {
      layoutManager = LinearLayoutManager(activity)
      itemAnimator = DefaultItemAnimator()
      adapter = postAdapter
    }

    noDataTextView = root.findViewById(R.id.tv_post_empty)

    fingerprintSwitch = root.findViewById<Switch>(R.id.sw_post_fingerprint).apply {
      setOnCheckedChangeListener { _, isChecked ->
        Constants.enableFingerprintLogin = isChecked
      }
    }

    return root
  }

  override fun onResume() {
    super.onResume()

    presenter.start()
  }

  override fun showPosts(posts: List<Post>) {
    postAdapter.replacePostList(posts)

    recyclerView.visibility = View.VISIBLE
    noDataTextView.visibility = View.GONE
  }

  override fun showNoData() {
    recyclerView.visibility = View.GONE
    noDataTextView.visibility = View.VISIBLE
  }

  override fun toPostDetailView(postId: Int) {
    val bundle = Bundle()
    bundle.putInt(EXTRA_POST_ID, postId)
    ActivityUtils.startActivity(cls = PostDetailActivity::class.java, extras = bundle)
  }

  override fun setFingerprintSwitchStatus(
    isOn: Boolean,
    isEnabled: Boolean
  ) {
    fingerprintSwitch.isChecked = isOn
    fingerprintSwitch.isEnabled = isEnabled
    if (!isEnabled) fingerprintSwitch.text = getString(R.string.fingerprint_switch_not_supported)
  }

  override fun showDialogFragment(dialogFragment: DialogFragment) {
    dialogFragment.show(fragmentManager, "TestTest")
  }

  inner class PostHolder internal constructor(
    inflater: LayoutInflater,
    val parent: ViewGroup
  ) : ViewHolder(inflater.inflate(R.layout.post_recycler_item_fragment, parent, false)) {

    private lateinit var post: Post

    private val titleTextView: TextView = itemView.findViewById(R.id.tv_post_item_title)
    private val contentTextView: TextView = itemView.findViewById(R.id.tv_post_item_content)

    init {
      itemView.setOnClickListener({
        toPostDetailView(post.postId)
      })
    }

    fun binding(post: Post) {
      this.post = post

      titleTextView.text = this.post.title
      contentTextView.text = this.post.content
    }
  }

  inner class PostAdapter constructor(private var postList: List<Post>) : Adapter<PostHolder>() {

    fun replacePostList(postList: List<Post>) {
      this.postList = postList
      notifyDataSetChanged()
    }

    override fun onCreateViewHolder(
      parent: ViewGroup,
      viewType: Int
    ): PostHolder {
      val inflater: LayoutInflater = LayoutInflater.from(activity)
      return PostHolder(inflater, parent)
    }

    override fun onBindViewHolder(
      holder: PostHolder,
      position: Int
    ) {
      val post = postList[position]
      holder.binding(post)
    }

    override fun getItemCount(): Int = postList.size
  }
}
